File: src/api/assumptions.ts
import { apiClient } from './client';
import type { Assumptions } from '@/types/assumptions';
import type { UnderwritingOutputs } from '@/types/outputs';

/* eslint-disable @typescript-eslint/no-explicit-any */

/** Map backend snake_case assumptions to frontend camelCase */
export function mapAssumptions(raw: any): Assumptions {
  if (!raw) {
    return {
      purchasePrice: 0,
      closingCosts: 0,
      renovationCost: 0,
      exitCapRate: 0,
      holdPeriod: 0,
      rentGrowth: 0,
      vacancy: 0,
      opexRatio: 0,
      ltv: 0,
      interestRate: 0,
      stabilizationMonths: 0,
    };
  }

  const purchasePrice = raw.purchase_price || 0;
  const loanAmount = raw.loan_amount || 0;

  return {
    purchasePrice,
    closingCosts: raw.closing_costs_pct != null ? raw.closing_costs_pct * 100 : 0,
    renovationCost: raw.renovation_cost || 0,
    exitCapRate: raw.exit_cap_rate != null ? raw.exit_cap_rate * 100 : 0,
    holdPeriod: raw.hold_period_years || 0,
    rentGrowth: raw.rent_growth_pct != null ? raw.rent_growth_pct * 100 : 0,
    vacancy: raw.vacancy_pct != null ? raw.vacancy_pct * 100 : 0,
    opexRatio: raw.opex_pct != null ? raw.opex_pct * 100 : 0,
    ltv: purchasePrice > 0 ? (loanAmount / purchasePrice) * 100 : 0,
    interestRate: raw.interest_rate != null ? raw.interest_rate * 100 : 0,
    stabilizationMonths: 0,
  };
}

/** Map frontend camelCase assumptions to backend snake_case */
function unmapAssumptions(data: Assumptions): Record<string, unknown> {
  const purchasePrice = data.purchasePrice || 0;
  return {
    purchase_price: purchasePrice,
    closing_costs_pct: (data.closingCosts || 0) / 100,
    renovation_cost: data.renovationCost || 0,
    exit_cap_rate: (data.exitCapRate || 0) / 100,
    hold_period_years: data.holdPeriod || 0,
    rent_growth_pct: (data.rentGrowth || 0) / 100,
    vacancy_pct: (data.vacancy || 0) / 100,
    opex_pct: (data.opexRatio || 0) / 100,
    loan_amount: purchasePrice * ((data.ltv || 0) / 100),
    interest_rate: (data.interestRate || 0) / 100,
    loan_term_years: data.holdPeriod || 10,
  };
}

/** Map backend snake_case outputs to frontend camelCase */
function mapOutputs(raw: any): UnderwritingOutputs {
  if (!raw) {
    return {
      irr: 0,
      equityMultiple: 0,
      cashOnCash: 0,
      noiGrowth: 0,
      requiredEquity: 0,
      breakEvenOccupancy: 0,
      projections: [],
      scenarios: [],
    };
  }

  return {
    irr: raw.irr != null ? raw.irr * 100 : 0,
    equityMultiple: raw.equity_multiple || 0,
    cashOnCash: raw.cash_on_cash != null ? raw.cash_on_cash * 100 : 0,
    noiGrowth: 0,
    requiredEquity: raw.equity_required || 0,
    breakEvenOccupancy: 0,
    projections: [],
    scenarios: [],
  };
}

/* eslint-enable @typescript-eslint/no-explicit-any */

export const assumptionsApi = {
  get: async (dealId: string): Promise<Assumptions> => {
    const raw = await apiClient.get<any>(`/deals/${dealId}/assumptions`);
    return mapAssumptions(raw);
  },

  update: async (dealId: string, data: Assumptions): Promise<Assumptions> => {
    const body = unmapAssumptions(data);
    const raw = await apiClient.put<any>(`/deals/${dealId}/assumptions`, body);
    return mapAssumptions(raw);
  },

  getOutputs: async (dealId: string): Promise<UnderwritingOutputs> => {
    const raw = await apiClient.get<any>(`/deals/${dealId}/outputs`);
    return mapOutputs(raw);
  },

  recalculate: async (dealId: string): Promise<UnderwritingOutputs> => {
    const raw = await apiClient.post<any>(`/deals/${dealId}/recalculate`);
    return mapOutputs(raw);
  },
};

File: app/schemas/assumptions.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class AssumptionsCreate(BaseModel):
    purchase_price: Optional[float] = Field(None, ge=0)
    closing_costs_pct: Optional[float] = Field(0.02, ge=0, le=1)
    renovation_cost: Optional[float] = Field(0.0, ge=0)
    loan_amount: Optional[float] = Field(None, ge=0)
    interest_rate: Optional[float] = Field(0.065, ge=0, le=1)
    loan_term_years: Optional[int] = Field(30, ge=1, le=50)
    gross_rental_income: Optional[float] = Field(None, ge=0)
    vacancy_pct: Optional[float] = Field(0.05, ge=0, le=1)
    opex_pct: Optional[float] = Field(0.35, ge=0, le=1)
    cap_rate_market: Optional[float] = Field(0.06, ge=0, le=1)
    exit_cap_rate: Optional[float] = Field(0.065, ge=0, le=1)
    hold_period_years: Optional[int] = Field(5, ge=1, le=30)
    rent_growth_pct: Optional[float] = Field(0.03, ge=-0.5, le=1)


class AssumptionsUpdate(BaseModel):
    purchase_price: Optional[float] = Field(None, ge=0)
    closing_costs_pct: Optional[float] = Field(None, ge=0, le=1)
    renovation_cost: Optional[float] = Field(None, ge=0)
    loan_amount: Optional[float] = Field(None, ge=0)
    interest_rate: Optional[float] = Field(None, ge=0, le=1)
    loan_term_years: Optional[int] = Field(None, ge=1, le=50)
    gross_rental_income: Optional[float] = Field(None, ge=0)
    vacancy_pct: Optional[float] = Field(None, ge=0, le=1)
    opex_pct: Optional[float] = Field(None, ge=0, le=1)
    cap_rate_market: Optional[float] = Field(None, ge=0, le=1)
    exit_cap_rate: Optional[float] = Field(None, ge=0, le=1)
    hold_period_years: Optional[int] = Field(None, ge=1, le=30)
    rent_growth_pct: Optional[float] = Field(None, ge=-0.5, le=1)


class AssumptionsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    deal_id: str
    purchase_price: Optional[float] = None
    closing_costs_pct: Optional[float] = None
    renovation_cost: Optional[float] = None
    loan_amount: Optional[float] = None
    interest_rate: Optional[float] = None
    loan_term_years: Optional[int] = None
    gross_rental_income: Optional[float] = None
    vacancy_pct: Optional[float] = None
    opex_pct: Optional[float] = None
    cap_rate_market: Optional[float] = None
    exit_cap_rate: Optional[float] = None
    hold_period_years: Optional[int] = None
    rent_growth_pct: Optional[float] = None
    created_at: datetime
    updated_at: datetime
